File: Problems/No.797_All_Paths_From_Source_to_Target/all_paths_from_source_to_target.cpp
class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        
        // parameters
        int target = graph.size()-1;
        std::vector<std::vector<int>> AllPaths2Target;
        std::queue<std::vector<int>> AllPaths;

        // start computing
        AllPaths.push({0});
        while (!AllPaths.empty()) {

            auto currPath = AllPaths.front();
            AllPaths.pop();

            for (auto nextNode: graph[currPath.back()]) {

                currPath.push_back(nextNode);
                if (nextNode == target) AllPaths2Target.push_back(currPath);
                else AllPaths.push(currPath);
                currPath.pop_back();

            }

        }

        return AllPaths2Target;

    }
};

File: Problems/No.797_All_Paths_From_Source_to_Target/all_paths_from_source_to_target.py
class Solution(object):
    def allPathsSourceTarget(self, graph):
        """
        :type graph: List[List[int]]
        :rtype: List[List[int]]
        """

        # parameters
        target = len(graph) - 1
        AllPaths2Target = []
        AllPaths = []

        # start computing
        AllPaths.append([0])
        while AllPaths:
            currPath = AllPaths.pop(0)
            for nextNode in graph[currPath[-1]]:
                currPath.append(nextNode)
                if (nextNode == target):
                    AllPaths2Target.append(list(currPath))
                else:
                    AllPaths.append(list(currPath))
                currPath.pop()
        
        return AllPaths2Target
